Src/userkybd.cpp

```diff
@@ -44,37 +45,244 @@ static void DrawBorder(HDC hDC, RECT rect, BOOL Depressed);
+		"Move stick for unshifted press...",
+		"Move stick for shifted press...",
+		"Move stick to unassign..."
+	}
+};
+
+// Table of BBC keys
+const BBCKey BBCKeys[] = {
+	// Unassigned must go first
+	{ IDK_UNASSIGN, "NONE", -9, 0 },
+
+	// Letter keys
+	{ IDK_A, "A", 4, 1 },
+	{ IDK_B, "B", 6, 4 },
+	{ IDK_C, "C", 5, 2 },
+	{ IDK_D, "D", 3, 2 },
+	{ IDK_E, "E", 2, 2 },
+	{ IDK_F, "F", 4, 3 },
+	{ IDK_G, "G", 5, 3 },
+	{ IDK_H, "H", 5, 4 },
+	{ IDK_I, "I", 2, 5 },
+	{ IDK_J, "J", 4, 5 },
+	{ IDK_K, "K", 4, 6 },
+	{ IDK_L, "L", 5, 6 },
+	{ IDK_M, "M", 6, 5 },
+	{ IDK_N, "N", 5, 5 },
+	{ IDK_O, "O", 3, 6 },
+	{ IDK_P, "P", 3, 7 },
+	{ IDK_Q, "Q", 1, 0 },
+	{ IDK_R, "R", 3, 3 },
+	{ IDK_S, "S", 5, 1 },
+	{ IDK_T, "T", 2, 3 },
+	{ IDK_U, "U", 3, 5 },
+	{ IDK_V, "V", 6, 3 },
+	{ IDK_W, "W", 2, 1 },
+	{ IDK_X, "X", 4, 2 },
+	{ IDK_Y, "Y", 4, 4 },
+	{ IDK_Z, "Z", 6, 1 },
+
+	// Number keys
+	{ IDK_0, "0", 2, 7 },
+	{ IDK_1, "1", 3, 0 },
+	{ IDK_2, "2", 3, 1 },
+	{ IDK_3, "3", 1, 1 },
+	{ IDK_4, "4", 1, 2 },
+	{ IDK_5, "5", 1, 3 },
+	{ IDK_6, "6", 3, 4 },
+	{ IDK_7, "7", 2, 4 },
+	{ IDK_8, "8", 1, 5 },
+	{ IDK_9, "9", 2, 6 },
+
+	// Function keys.
+	{ IDK_F0, "F0", 2, 0 },
+	{ IDK_F1, "F1", 7, 1 },
+	{ IDK_F2, "F2", 7, 2 },
+	{ IDK_F3, "F3", 7, 3 },
+	{ IDK_F4, "F4", 1, 4 },
+	{ IDK_F5, "F5", 7, 4 },
+	{ IDK_F6, "F6", 7, 5 },
+	{ IDK_F7, "F7", 1, 6 },
+	{ IDK_F8, "F8", 7, 6 },
+	{ IDK_F9, "F9", 7, 7 },
+
+	// Special keys.
+	{ IDK_LEFT, "LEFT", 1, 9 },
+	{ IDK_RIGHT, "RIGHT", 7, 9 },
+	{ IDK_UP, "UP", 3, 9 },
+	{ IDK_DOWN, "DOWN", 2, 9 },
+	{ IDK_BREAK, "BREAK", -2, -2 },
+	{ IDK_COPY, "COPY", 6, 9 },
+	{ IDK_DEL, "DELETE", 5, 9 },
+	{ IDK_CAPS, "CAPS-LOCK", 4, 0 },
+	{ IDK_TAB, "TAB", 6, 0 },
+	{ IDK_CTRL, "CTRL", 0, 1 },
+	{ IDK_SPACE, "SPACE", 6, 2 },
+	{ IDK_RETURN, "RETURN", 4, 9 },
+	{ IDK_ESC, "ESCAPE", 7, 0 },
+	{ IDK_SHIFT_L, "SHIFT", 0, 0 },
+	{ IDK_SHIFT_R, "SHIFT", 0, 0 },
+	{ IDK_SHIFT_LOCK, "SHIFT-LOCK", 5, 0 },
+
+	// Special Character keys.
+	{ IDK_SEMI_COLON, ";", 5, 7 },
+	{ IDK_EQUALS, "=", 1, 7 },
+	{ IDK_COMMA, ",", 6, 6 },
+	{ IDK_CARET, "^", 1, 8 },
+	{ IDK_DOT, ".", 6, 7 },
+	{ IDK_FWDSLASH, "/", 6, 8 },
+	{ IDK_STAR, "*", 4, 8 },
+	{ IDK_OPEN_SQUARE, "[", 3, 8 },
+	{ IDK_BACKSLASH, "\\", 7, 8 },
+	{ IDK_CLOSE_SQUARE, "]", 5, 8 },
+	{ IDK_AT, "@", 4, 7 },
+	{ IDK_UNDERSCORE, "_", 2, 8 },
+};
+
+const std::pair<const char*, const char*> BBCKeyAliases[] = {
+	{ "DEL", "DELETE" },
+	{ "CAPS", "CAPS-LOCK" },
+	{ "CAPSLOCK", "CAPS-LOCK" },
+	{ "CONTROL", "CTRL" },
+	{ "ESC", "ESCAPE" },
+	{ "SHIFTLOCK", "SHIFT-LOCK" }
 };
 
+// Number of inputs assigned to each key (for joystick dialog highlight)
+static int assignedKeysCount[_countof(BBCKeys)] = {};
+
+/****************************************************************************/
+
+const BBCKey& GetBBCKeyByResId(int ctrlId)
+{
+    using resIdToKeyMapType = std::map<int, const BBCKey*>;
+
+    // Construct map on first use by lambda
+    static resIdToKeyMapType resIdToKeyMap = []()
```

Comment:
Can this map (and the similar ones in other functions) be declared `const`?

Reply:
Definitely